doc/design.md

```diff
@@ -0,0 +1,949 @@
+
+
+### <a name="design-details-copy-on-write-mechanism"></a>Copy-on-write mechanism
+
+##### Context
+
+The copy-on-write mechanism is hidden in the API and simply handled internally in `Variable`:
+A copy-on-write pointer hold a `VariableConcept`.
+To maintain sharing we only require code to be `const`-correct.
+*Question: Cannot use `const` in Python, need methods/tags with different names that explicitly request mutable or immutable access. What is the bast naming convention?*
+
+##### Rationale
+
+The copy-on-write mechanism is something that has since long been adopted in Mantid, e.g., in `DataObjects::Workspace2D` for the histogram data and for `Geometry::DetectorInfo`.
+However, for the interface of `API::MatrixWorkspace` (and in connection to that `HistogramData::Histogram`) this has led to an interface that is too complicated, with different API functions for read-only access, write-access, and access for sharing.
+
+
+### <a name="design-details-uncertainties-are-stored-as-variances"></a>Uncertainties are stored as variances, not standard deviations
+
+##### Rationale
+
+Mantid currently stores uncertainties as standard deviation (except for `DataObjects::MDHistoWorkspace`).
+However, most computations require the variance for propagation of uncertainties.
+To this end it is often squared in place (breaking assumptions about, e.g., `HistogramData::Histogram` and `DataObjects::Workspace2D`, and subsequently converted back to standard deviation by taking the square root.
+This also adds a little computational overhead.
+Having a clear and well defined strategy will reduce the risk of confusion.
+
+##### Implementation
+
+The go-to tag for storing uncertainties will be `Data::Variance`.
+`Data::StdDev` will exist, but is for *access* only, i.e., attempting to create a `Variable` with that tag should fail.
+Instead `Data::StdDev` can by used in combination with `DatasetView` to *read* the standard deviations if required, e.g., for visualization purposes.
+If it turns out to be useful it would also be possible to provide setters --- as opposed to a simple assignment which is possible if the actual underlying data is referenced --- for standard deviation using `DatasetView` .
+
+
+### <a name="design-details-distribution-histogram-distinguised-by-unit"></a>"Distributions" and "histogram data" are distinguished by their unit
+
+##### Rationale
+
+Currently `API::MatrixWorkspace` (and by extension `HistogramData::Histogram`) keep a flag indicating whether data in a histogram has been divided by its bin width.
+A considerable number of algorithms are checking this manually, branching to different code sections.
+`HistogramData::Histogram` attempted to solve this by introducing getters like `Histogram::counts()` and `Histogram::frequencies()`.
+However, it seems to be useful only in a rather limited number of places, whereas it adds a significant amount of complication to the API.
+The flag adds complication and should thus be removed.
```

Comment:
I did not like the inconsistent flag, and am a fan of the new/curent approach. For writing client code:
```
f = hisotgram.frequencies()
```
is much better than what I think you are now proposing:
1. Query units
2. If correct use value, otherwise apply conversion yourself.
Let me know if I'm not understanding that correctly.

Reply:
Yes, that is what I am proposing (of course there could be helpers). I agree that writing `histogram.frequencies()` has advantages, but:

- It returns by value, i.e., needs to make a copy of data in many cases, and more importantly it cannot be written to.
- Many or most algorithms actually do not seem to care whether data is frequencies or counts, so we have duplication in the API to also support the "don't care" getter.
- Need to explain yet another concept to developers.

Nothing keeps us from providing converters and read-only views for convenience but I think this should not be part of the data type.